strawberryfields/program.py

```diff
@@ -464,6 +469,49 @@ def assert_number_of_modes(self, device):
                 f"only supports a {device.modes}-mode program."
             )
 
+    def assert_number_of_measurements(self, device):
+        """Check that the number of measurements in the circuit doesn't exceed the number of allowed
+        measurements according to the device specification.
+
+        Args:
+            device (~strawberryfields.api.DeviceSpec): Device specification object to use.
+                ``device.modes`` must be a dictionary, containing the maximum number of allowed
+                measurements for the specified target.
+
+        """
+        num_pnr, num_homodyne, num_heterodyne = 0, 0, 0
+
+        try:
+            max_pnr = device.modes["max"]["pnr"]
+            max_homodyne = device.modes["max"]["homodyne"]
+            max_heterodyne = device.modes["max"]["heterodyne"]
+        except (KeyError, TypeError) as e:
+            raise KeyError(
+                "Device specification must contain an entry for the maximum allowed number "
+                "of measurments. Have you specified the correct target?"
+            ) from e
+
+        for c in self.circuit:
+            op_name = str(c.op)
+            if "MeasureFock" in op_name:
+                num_pnr += len(c.reg)
+            if "MeasureHomodyne" in op_name or "MeasureX" in op_name or "MeasureP" in op_name:
+                num_homodyne += len(c.reg)
+            if "MeasureHeterodyne" in op_name or "MeasureHD"in op_name:
+                num_heterodyne += len(c.reg)
+
+        too_many_measurments = (
+            num_pnr > max_pnr, num_homodyne > max_homodyne, num_heterodyne > max_heterodyne
+        )
+        if any(too_many_measurments):
+            raise CircuitError(
+                f"This program contains {num_pnr} fock measurements, {num_homodyne} homodyne "
+                f"measurements and {num_heterodyne} heterodyne measurements.\n"
+                f"A maximum of {max_pnr} fock measurements, {max_homodyne} homodyne "
+                f"measurements, {max_heterodyne} heterodyne measurements are supported."
```

Comment:
Would we want to have a separate error for each measurement type? It's inexpensive for the user to check & update the code because this is still caught locally.

Reply:
True, could fix that. I don't mind it either way tbh.